solr/core/src/java/org/apache/solr/security/AllowListUrlChecker.java

```diff
@@ -0,0 +1,201 @@
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+                    "Invalid URL syntax in '" + URL_ALLOW_LIST + "' configuration: " + config.getAllowUrls(), e);
+        }
+    }
+
+    /**
+     * @see #checkAllowList(List, ClusterState)
+     */
+    public void checkAllowList(List<String> urls) throws MalformedURLException {
+        checkAllowList(urls, null);
+    }
+
+    /**
+     * Checks that the given URLs are present in the configured allow-list or in the provided {@link ClusterState}
+     * (in case of cloud mode).
+     *
+     * @param urls         The list of urls to check.
+     * @param clusterState The up to date {@link ClusterState}, can be null in case of non-cloud mode.
+     * @throws MalformedURLException If an URL is invalid.
+     * @throws SolrException         If an URL is not present in the allow-list or in the provided {@link ClusterState}.
+     */
+    public void checkAllowList(List<String> urls, @Nullable ClusterState clusterState) throws MalformedURLException {
+        Set<String> clusterHostAllowList = clusterState == null ? Collections.emptySet() : clusterState.getHostAllowList();
+        for (String url : urls) {
+            String hostPort = parseHostPort(url);
+            if (!clusterHostAllowList.contains(hostPort) && !hostAllowList.contains(hostPort)) {
+                throw new SolrException(SolrException.ErrorCode.FORBIDDEN, "URL " + url +
+                        " is neither a live node of the cluster nor in the configured '" +
+                        URL_ALLOW_LIST + "' " + hostAllowList);
+            }
+        }
+    }
+
+    /**
+     * Whether this checker has been created with a non-empty allow-list of URLs.
+     */
+    public boolean hasExplicitAllowList() {
+        return !hostAllowList.isEmpty();
+    }
+
+    /**
+     * Whether the URL checking is enabled. Only {@link #ALLOW_ALL} returns false.
+     */
+    public boolean isEnabled() {
+        return true;
+    }
+
+    /**
+     * Only for testing.
+     */
+    @VisibleForTesting
+    public Set<String> getHostAllowList() {
+        return hostAllowList == null ? null : Collections.unmodifiableSet(hostAllowList);
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + " [allowList=" + hostAllowList + "]";
+    }
+
+    @VisibleForTesting
+    static Set<String> parseHostPorts(List<String> urls) throws MalformedURLException {
+        if (urls == null || urls.isEmpty()) {
+            return Collections.emptySet();
+        }
+        Set<String> hostPorts = new HashSet<>((int) (urls.size() / 0.7f) + 1);
+        for (String urlString : urls) {
+            hostPorts.add(parseHostPort(urlString));
+        }
+        return hostPorts;
+    }
+
+    private static String parseHostPort(String url) throws MalformedURLException {
+        url = url.trim();
+        URL u;
+        if (!url.startsWith("http://") && !url.startsWith("https://")) {
```

Comment:
Are there circumstances where we would be looking at a different protocol? Since we're using this for more than just shards now, maybe this will be part of BackupRestore and will need to accept `s3://` URLs?

Reply:
+1